src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp

```diff
@@ -0,0 +1,202 @@
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Reduction.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Event.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace Frame {
+struct Inertial;
+}  // namespace Frame
+/// \endcond
+
+namespace dg {
+namespace Events {
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors,
+          typename EventRegistrars>
+class ObserveVolumeIntegrals;
+
+namespace Registrars {
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors>
+// Presence of size_t template argument requires to define this struct
+// instead of using Registration::Registrar alias.
+struct ObserveVolumeIntegrals {
+  template <typename RegistrarList>
+  using f = Events::ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
+                                           Tensors, RegistrarList>;
+};
+}  // namespace Registrars
+
+template <size_t VolumeDim, typename ObservationValueTag, typename Tensors,
+          typename EventRegistrars =
+              tmpl::list<Registrars::ObserveVolumeIntegrals<
+                  VolumeDim, ObservationValueTag, Tensors>>>
+class ObserveVolumeIntegrals;
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief %Observe the volume integrals of the tensors over the domain.
+ *
+ * Writes reduction quantities:
+ * - `ObservationValueTag`
+ * - `Volume` = volume of the domain
+ * - `VolInt(*)` = volume integral of the tensor
+ *
+ * \warning Currently, only one reduction observation event can be
+ * triggered at a given observation value.  Causing multiple events to run at
+ * once will produce unpredictable results.
+ */
+template <size_t VolumeDim, typename ObservationValueTag, typename... Tensors,
+          typename EventRegistrars>
+class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
+                             tmpl::list<Tensors...>, EventRegistrars>
+    : public Event<EventRegistrars> {
+ private:
+  struct VectorPlus {
+    std::vector<double> operator()(const std::vector<double>& lhs,
+                                   const std::vector<double>& rhs) const
+        noexcept {
+      ASSERT(lhs.size() == rhs.size(),
+             "Error: vector sizes in `VectorPlus` operator do not match.");
+      std::vector<double> result(lhs.size());
+      for (size_t i = 0; i < lhs.size(); ++i) {
+        result[i] = lhs[i] + rhs[i];
+      }
+      return result;
+    }
+  };
+
+  using VolIntDatum = Parallel::ReductionDatum<std::vector<double>, VectorPlus>;
+
+  using ReductionData = tmpl::wrap<
+      tmpl::list<Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
+                 Parallel::ReductionDatum<double, funcl::Plus<>>, VolIntDatum>,
+      Parallel::ReductionData>;
```

Comment:
I think we should move this to runtime because if we want to control different norm types and variables via the input file it'll be better to have more runtime control.

Reply:
yep that sounds like the right thing to do 👍